src/components/tl/ucp/recv/recv.c

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates. 2022.
+ *
+ * See file LICENSE for terms.
+ */
+#include "config.h"
+#include "tl_ucp.h"
+#include "tl_ucp_coll.h"
+#include "tl_ucp_sendrecv.h"
+#include "recv.h"
+
+ucc_status_t ucc_tl_ucp_recv_init(ucc_tl_ucp_task_t *task)
+{
+    task->super.post     = ucc_tl_ucp_recv_start;
+    task->super.progress = ucc_tl_ucp_recv_progress;
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_recv_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t   *task       = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t   *team       = TASK_TEAM(task);
+    ucc_coll_args_t     *args       = &TASK_ARGS(task);
+    uint64_t             peer       = args->root;
+    void                *rbuf       = args->dst.info.buffer;
+    ucc_memory_type_t    mem_type   = args->dst.info.mem_type;
+    size_t               count      = args->dst.info.count;
+    ucc_datatype_t       dt         = args->dst.info.datatype;
+    size_t               dt_size    = ucc_dt_size(dt);
+    size_t               data_size  = count * dt_size;
+
+    UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(rbuf, data_size, mem_type, peer, team, task), task, out);
+    if (task->tagged.recv_posted == task->tagged.recv_completed) {
```

Comment:
I think this if statement is not needed, you can just do return ucc_progress_queue_enqueue(...)